tests/interlocked_win32_ut/interlocked_win32_ut.c

```diff
@@ -0,0 +1,708 @@
+}
+
+/*Tests_SRS_INTERLOCKED_WIN32_43_007: [interlocked_and_64 shall call InterlockedAnd64 from windows.h.]*/
+/*Tests_SRS_INTERLOCKED_WIN32_43_008 : [interlocked_and_64 shall return the initial value of * destination.]*/
+TEST_FUNCTION(interlocked_and_64_calls_InterlockedAnd64)
+{
+    ///arrange
+    volatile uint64_t destination = 0xF0F0F0F0F0F0F0F0;
+    uint64_t value = 0x0F0F0F0F0F0F0F0F;
+    STRICT_EXPECTED_CALL(mock_InterlockedAnd64((volatile LONG64*)&destination, (LONG64)value))
+        .SetReturn(0xF0F0F0F0F0F0F0F0);
+    ///act
+    uint64_t return_val = (uint64_t)interlocked_and_64((volatile int64_t*)&destination, (int64_t)value);
+
+    ///assert
+    ASSERT_ARE_EQUAL(char_ptr, umock_c_get_expected_calls(), umock_c_get_actual_calls());
+    ASSERT_ARE_EQUAL(uint64_t, 0xF0F0F0F0F0F0F0F0, return_val);
+}
+
+/*Tests_SRS_INTERLOCKED_WIN32_43_009: [interlocked_and_8 shall call InterlockedAnd8 from windows.h.]*/
+/*Tests_SRS_INTERLOCKED_WIN32_43_010 : [interlocked_and_8 shall return the initial value of * destination.]*/
+TEST_FUNCTION(interlocked_and_8_calls_InterlockedAdd8)
+{
+    ///arrange
+    volatile uint8_t destination = 0xF0;
+    uint8_t value = 0x0F;
+    STRICT_EXPECTED_CALL(mock_InterlockedAnd8((volatile char*)&destination, (char)value))
+        .SetReturn(0xF0);
+    ///act
+    uint8_t return_val = (uint8_t)interlocked_and_8((volatile int8_t*)&destination, (int8_t)value);
+
+    ///assert
+    ASSERT_ARE_EQUAL(char_ptr, umock_c_get_expected_calls(), umock_c_get_actual_calls());
+    ASSERT_ARE_EQUAL(uint8_t, 0xF0, return_val);
+}
+
+/*Tests_SRS_INTERLOCKED_WIN32_43_011: [interlocked_compare_exchange shall call InterlockedCompareExchange from windows.h.]*/
+/*Tests_SRS_INTERLOCKED_WIN32_43_012 : [interlocked_compare_exchange shall return the initial value of * destination.]*/
+TEST_FUNCTION(interlocked_compare_exchange_calls_InterlockedCompareExchange)
+{
+    ///arrange
+    volatile int32_t destination = INT32_MAX;
+    int32_t comperand = INT32_MAX;
+    int32_t exchange = INT32_MIN;
+    STRICT_EXPECTED_CALL(mock_InterlockedCompareExchange((volatile LONG*)&destination, (LONG)exchange, (LONG)comperand))
+        .SetReturn(INT32_MAX);
+
+    ///act
+    int32_t return_val = interlocked_compare_exchange(&destination, exchange, comperand);
+
+    ///assert
+    ASSERT_ARE_EQUAL(char_ptr, umock_c_get_expected_calls(), umock_c_get_actual_calls());
+    ASSERT_ARE_EQUAL(int32_t, INT32_MAX, return_val);
+}
+
+/*Tests_SRS_INTERLOCKED_WIN32_43_013: [interlocked_compare_exchange_128 shall call InterlockedCompareExchange128 from windows.h.]*/
+/*Tests_SRS_INTERLOCKED_WIN32_43_014 : [interlocked_compare_exchange_128 shall return true if* comperand_result equals the original value of* destination.]*/
+TEST_FUNCTION(interlocked_compare_exchange_128_calls_InterlockedCompareExchange128_true_case)
+{
+    ///arrange
+    volatile int64_t* destination = (volatile int64_t*)_aligned_malloc(2 * sizeof(int64_t), 16);
```

Comment:
>destination [](start = 22, length = 11)

destination[2] is equally good.

Reply:
If I declare it as destination[2] I must provide a brace enclosed initializer list. Creating the array on the stack gives an alignment error in InterlockedCompareExchange128.

---
In reply to: [427743989](https://github.com/Azure/azure-c-util/pull/43#discussion_r427743989) [](ancestors = 427743989)